src/Serialization/Address.js

```diff
@@ -0,0 +1,148 @@
+    const ret = null;
+    try{
+        const addr = CardanoWasm.Address.from_bytes(bytes);
+        ret = CardanoWasm.BaseAddress.from_address(addr);
+    }
+    catch(e){
+        console.log(e);
+    }
+    if (ret==null){
+        return maybe.nothing;
+    }
+    return maybe.just(ret);
+};
+
+exports.rewardAddressFromBytesImpl = maybe => bytes => {
+    const ret = null;
+    try {
+        const addr = CardanoWasm.Address.from_bytes(bytes);
+        ret = CardanoWasm.RewardAddress.from_address(addr);
+    }
+    catch (e) {
+        console.log(e);
+    }
+    if (ret == null) {
+        return maybe.nothing;
+    }
+    return maybe.just(ret);
+};
+
+exports.baseAddressFromBech32Impl = maybe => str => {
+    const ret = null;
+    try {
+        const addr = CardanoWasm.Address.from_bech32(str);
+        ret = CardanoWasm.BaseAddress.from_address(addr);
+    }
+    catch (e) {
+        console.log(e);
+    }
+    if (ret == null) {
+        return maybe.nothing;
+    }
+    return maybe.just(ret);
+};
+
+exports.rewardAddressFromBech32Impl = maybe => str => {
+    const ret = null;
+    try {
+        const addr = CardanoWasm.Address.from_bech32(str);
+        ret = CardanoWasm.RewardAddress.from_address(addr);
+    }
+    catch (e) {
+        console.log(e);
+    }
+    if (ret == null) {
+        return maybe.nothing;
+    }
+    return maybe.just(ret);
+};
+
+exports.headerCheckBaseAddr = maybe => checks => bytes => baseAddr => {
+    // NOTE from CIP-19 and CSL codebase:
+    // shelley payment addresses:
+    // bit 7: 0
+    // bit 6: base/other
+    // bit 5: pointer/enterprise [for base: stake cred is keyhash/scripthash]
+    // bit 4: payment cred is keyhash/scripthash
+    // bits 3-0: network id
+    //
+    // reward addresses:
+    // bits 7-5: 111
+    // bit 4: credential is keyhash/scripthash
+    // bits 3-0: network id
+    //
+    // byron addresses:
+    // bits 7-4: 1000
+    const paymcheck = maybe.from(null, checks.payment);
+    if (paymcheck != null && !checkBitSet(bytes, 4, paymcheck)){
+        return maybe.nothing;
+    }
+    const delegcheck = maybe.from(null, checks.delegation);
+    if (delegcheck != null && !checkBitSet(bytes, 3, delegcheck)) {
+        return maybe.nothing;
+    }
+    return maybe.just(baseAddr);
+};
+
+exports.headerCheckRewardAddr = maybe => checks => bytes => baseAddr => {
+    // NOTE from CIP-19 and CSL codebase:
+    // shelley payment addresses:
+    // bit 7: 0
+    // bit 6: base/other
+    // bit 5: pointer/enterprise [for base: stake cred is keyhash/scripthash]
+    // bit 4: payment cred is keyhash/scripthash
+    // bits 3-0: network id
+    //
+    // reward addresses:
+    // bits 7-5: 111
+    // bit 4: credential is keyhash/scripthash
+    // bits 3-0: network id
+    //
+    // byron addresses:
+    // bits 7-4: 1000
+    const paymcheck = maybe.from(null, checks.payment);
+    if (paymcheck != null && !checkBitSet(bytes, 4, paymcheck)) {
```

Comment:
What is `checkBitSet`?

Reply:
Oh I forgot about it. It will check if n'th bit is as expected